File: app/src/main/java/workshop1024/com/xproject/controller/fragment/home/FilterFragment.java
package workshop1024.com.xproject.controller.fragment.home;

import android.support.design.widget.Snackbar;

import java.util.ArrayList;
import java.util.List;

import workshop1024.com.xproject.R;
import workshop1024.com.xproject.controller.adapter.HomeSubListAdapter;
import workshop1024.com.xproject.controller.fragment.home.news.FilterNewsFragment;
import workshop1024.com.xproject.model.subinfo.SubInfo;

/**
 * 抽屉导航HomeFragment的子Frament HomeFragment的ViewPager的子Fragment，按照过滤关键字分类展示
 */
public class FilterFragment extends HomeSubFragment {

    public static FilterFragment newInstance() {
        FilterFragment filterFragment = new FilterFragment();
        return filterFragment;
    }

    @Override
    public void onRefresh() {
        refreshFilterList();
    }

    @Override
    protected void loadData() {
        refreshFilterList();
    }

    /**
     * 刷新过滤器列表
     */
    private void refreshFilterList() {
        Snackbar.make(mHomesubFragmentBinding.getRoot(), "Fetch more filters ...", Snackbar.LENGTH_SHORT).show();
        mHomeSubFragmentHanlders.isRefreshing.set(true);
        mSubInfoRepository.getFilterSubInfos(this);
    }

    @Override
    public void onSubInfosLoaded(List<SubInfo> subInfoList) {
        if (mIsForeground) {
            mSubInfoList = subInfoList;
            mHomeSubFragmentHanlders.isRefreshing.set(false);
            HomeSubListAdapter homeSubListAdapter = new HomeSubListAdapter(subInfoList, this);
            mHomesubFragmentBinding.homesubRecyclerviewList.setAdapter(homeSubListAdapter);
            Snackbar.make(mHomesubFragmentBinding.getRoot(), "Fetch " + subInfoList.size() + " filters ...", Snackbar.LENGTH_SHORT).show();
        }
    }

    @Override
    public void onDataNotAvailable() {
        Snackbar.make(mHomesubFragmentBinding.getRoot(), "No filters refresh...", Snackbar.LENGTH_SHORT).show();
    }

    @Override
    public void onSubListItemClick(SubInfo subInfo) {
        FilterNewsFragment filterNewsFragment = FilterNewsFragment.Companion.newInstance(subInfo.getInfoId());
        getActivity().getSupportFragmentManager().beginTransaction().replace(R.id.mainright_framelayout_fragments, filterNewsFragment)
                .addToBackStack("").commit();
        getActivity().setTitle(subInfo.getName());
    }

    @Override
    public void markAsRead() {
        List<String> subInfoIds = new ArrayList<>();
        for (SubInfo subInfo : mSubInfoList) {
            subInfoIds.add(subInfo.getInfoId());
        }

        mSubInfoRepository.markeFilterSubInfoesAsRead(subInfoIds);

        refreshFilterList();
    }
}

File: android/app/src/main/java/workshop1024/com/xproject/controller/fragment/home/FilterFragment.kt
package workshop1024.com.xproject.controller.fragment.home

import com.google.android.material.snackbar.Snackbar
import java.util.ArrayList

import workshop1024.com.xproject.R
import workshop1024.com.xproject.controller.adapter.HomeSubListAdapter
import workshop1024.com.xproject.controller.fragment.home.news.FilterNewsFragment
import workshop1024.com.xproject.model.subinfo.SubInfo

/**
 * 抽屉导航HomeFragment的子Frament HomeFragment的ViewPager的子Fragment，按照过滤关键字分类展示
 */
class FilterFragment : HomeSubFragment() {

    override fun onRefresh() {
        refreshFilterList()
    }

    override fun loadData() {
        refreshFilterList()
    }

    /**
     * 刷新过滤器列表
     */
    private fun refreshFilterList() {
        Snackbar.make(mHomesubFragmentBinding.root, "Fetch more filters ...", Snackbar.LENGTH_SHORT).show()
        mHomeSubFragmentHanlders.isRefreshing.set(true)
        mSubInfoRepository.getFilterSubInfos(this)
    }

    override fun onSubInfosLoaded(subInfoList: List<SubInfo>) {
        if (mIsForeground) {
            mSubInfoList = subInfoList
            mHomeSubFragmentHanlders.isRefreshing.set(false)
            val homeSubListAdapter = HomeSubListAdapter(subInfoList, this)
            mHomesubFragmentBinding.homesubRecyclerviewList.adapter = homeSubListAdapter
            Snackbar.make(mHomesubFragmentBinding.root, "Fetch " + subInfoList.size + " filters ...", Snackbar.LENGTH_SHORT).show()
        }
    }

    override fun onDataNotAvailable() {
        Snackbar.make(mHomesubFragmentBinding.root, "No filters refresh...", Snackbar.LENGTH_SHORT).show()
    }

    override fun onSubListItemClick(subInfo: SubInfo) {
        val filterNewsFragment = FilterNewsFragment.newInstance(subInfo.infoId)
        activity?.supportFragmentManager?.beginTransaction()?.replace(R.id.mainright_framelayout_fragments, filterNewsFragment)
                ?.addToBackStack("")?.commit()
        activity?.title = subInfo.name
    }

    public override fun markAsRead() {
        val subInfoIds = ArrayList<String>()
        for ((infoId) in mSubInfoList!!) {
            subInfoIds.add(infoId)
        }

        mSubInfoRepository.markeFilterSubInfoesAsRead(subInfoIds)

        refreshFilterList()
    }

    companion object {

        fun newInstance(): FilterFragment {
            return FilterFragment()
        }
    }
}
